torch/csrc/autograd/engine.cpp

```diff
@@ -747,18 +729,19 @@ variable_list Engine::execute_with_graph_task(
 
   // Not a worker
   if (worker_device == NO_DEVICE) {
-    // Wait for all tasks to complete
-    graph_task->not_done_.wait(
-        lock, [&graph_task] { return graph_task_completed(*graph_task); });
+    // graph_task_exec_post_processing is done when the Future is marked as
+    // completed in mark_graph_task_completed.
+    return graph_task->future_result_;
   } else {
     graph_task->owner_ = worker_device;
     ++total_depth;
-    if(current_depth >= max_recursion_depth_){
+    if (current_depth >= max_recursion_depth_) {
       // See Note [Reentrant backwards]
       // If reached the max depth, switch to a different thread
       add_thread_pool_task(graph_task);
-      graph_task->not_done_.wait(
-          lock, [&graph_task] { return graph_task_completed(*graph_task); });
+      // graph_task_exec_post_processing is done when the Future is marked as
+      // completed in mark_graph_task_completed.
+      return graph_task->future_result_;
```

Comment:
Actually, looking into the codebase, is `total_depth` used anywhere? We seem to maintain it across threads but we never use the value?

Reply:
It is used here to set `reentrant_depth_` in GraphTask: https://github.com/pytorch/pytorch/blob/877c96cddfebee00385307f9e1b1f3b4ec72bfdc/torch/csrc/autograd/engine.cpp#L709. `reentrant_depth_` is used to order items in the priority_queue.

Reply:
Ho right. My bad.
current depth is the depth of the current thread while total depth is the depth counting cases where you use the thread pool. SO you do want it for both codepath. If you move it as you suggest, it will just become the same as current depth.

Reply:
It won't be the same as current depth if I move it, the reason being is that the total_depth is passed into the GraphTask as reentrant_depth and in other threads the total_depth is bootstrapped from reentrant_depth here: https://github.com/pytorch/pytorch/blob/master/torch/csrc/autograd/engine.cpp#L375. Also, note that the incremented value of total_depth is never used in this codepath. The codepath only calls `add_thread_pool_task` which does not use total_depth.

Reply:
Right in this codepath, the value stored in the graphtask is used directly.
It is fine to increment it only where we use it I think. So the same place as current_depth.